### PR TITLE
Timer fix.

### PR DESCRIPTION
**Pull request by edwarnicke**

### CODE REVIEW

File: pkg/networkservice/common/timeout/server.go [hunk 1]
```diff
@@ -60,17 +59,20 @@ func NewServer(onTimout *networkservice.NetworkServiceServer) networkservice.Net
 }
 
 func (t *timeoutServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
-	ct, err := t.createTimer(ctx, request)
+	conn, err := next.Server(ctx).Request(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+	ct, err := t.createTimer(request.GetConnection())
```
Comment (edwarnicke) [inline pkg/networkservice/common/timeout/server.go:66]: As pointed out by @Bolodya1997 [here](https://github.com/networkservicemesh/sdk/pull/520#issue-502056572) we should be computing our timer based on the returned connection.

**Comment by Bolodya1997** [comment]
It is the only one small thing fixed by #520, another 2 **huge** problems are:
* timeout double closes part of the given chain: https://github.com/networkservicemesh/sdk/pull/520#issuecomment-708197650
* if timer fires with Close after the client sends new Request, but before the Request comes into the `timeout` chain element, there will be both Close and Request in parallel executing for the same connection ID

My fault, I have not updated the PR description, but now it is targeted to completely rework the `timeout` chain element and fix all of these problems.

**Review by edwarnicke — COMMENTED** [review]
> It is the only one small thing fixed by #520, another 2 **huge** problems are:
> 
> * timeout double closes part of the given chain: [#520 (comment)](https://github.com/networkservicemesh/sdk/pull/520#issuecomment-708197650)

So the issue mentioned in [#520 (comment)](https://github.com/networkservicemesh/sdk/pull/520#issuecomment-708197650) I do not believe is an issue with this solution, as https://github.com/networkservicemesh/sdk/blob/8f523a2e146549aa68edd50280628ab3b31d13fb/pkg/networkservice/common/timeout/server.go#L97-L107 no longer copies Values from the context of the original Request.

> * if timer fires with Close after the client sends new Request, but before the Request comes into the `timeout` chain element, there will be both Close and Request in parallel executing for the same connection ID
> 
> My fault, I have not updated the PR description, but now it is targeted to completely rework the `timeout` chain element and fix all of these problems.
> 
> **UPD:** created an issue #551.

**Comment by edwarnicke** [comment]
> * if timer fires with Close after the client sends new Request, but before the Request comes into the `timeout` chain element, there will be both Close and Request in parallel executing for the same connection ID

What was your solution to that problem?

**Comment by Bolodya1997** [comment]
> > * if timer fires with Close after the client sends new Request, but before the Request comes into the `timeout` chain element, there will be both Close and Request in parallel executing for the same connection ID
> 
> What was your solution to that problem?

I  use `MutexGroup` to serialize all incoming Close and Request events in the `timeout` chain element. If timer fires with Close, Request will be postponed for the rest of the chain until Close finishes (or fails).
This is another reason why I have stopped using `*onTimeout` - `timeout` can't make such thing if timer fires Close event not on `timeout` itself, but on some other chain element in the chain before, so there can possibly be some Request and Close walking in parallel **before** the `timeout` starts serializing them.

File: pkg/networkservice/common/timeout/server.go [hunk 2]
```diff
@@ -83,18 +87,22 @@ func (t *timeoutServer) Close(ctx context.Context, conn *networkservice.Connecti
...
+func (t *timeoutServer) createTimer(ctx context.Context, conn *networkservice.Connection) (*time.Timer, error) {
+	conn = conn.Clone()
+	expireTime, err := ptypes.Timestamp(conn.GetCurrentPathSegment().GetExpires())
 	if err != nil {
 		return nil, err
 	}
-	duration := time.Until(expireTime)
-	conn := request.GetConnection().Clone()
-	return time.AfterFunc(duration, func() {
-		newCtx := extend.WithValuesFromContext(context.Background(), ctx)
```
[the remark this answers is not in the document]
Comment (Bolodya1997) [reply pkg/networkservice/common/timeout/server.go:95]: Using `*onTimeout` breaks inner `timeout` per-Connection event serialization, please take a look https://github.com/networkservicemesh/sdk/pull/547#issuecomment-713289683.

**Comment by denis-tingaikin** [comment]
Rework of timers has been done in https://github.com/networkservicemesh/sdk/pull/520

So closing this.

**Pull request closed without merging**